src/experiments/tcpsocket/api.js

```diff
@@ -0,0 +1,181 @@
+        }
+    };
+
+    return {
+        /**
+         * Return a Promise that will be resolved with the next (non-data) event
+         * received by the socket.  If there are queued events, the Promise will
+         * be immediately resolved (but you won't see that until a future turn of
+         * the event loop).
+         */
+        waitForEvent() {
+            if (pendingResolve) {
+                throw new Error("only one wait allowed at a time.");
+            }
+
+            if (receivedEvents.length) {
+                return Promise.resolve(receivedEvents.shift());
+            }
+
+            console.log("(" + socketType + " waiting for event)\n");
+            return new Promise(function(resolve, reject) {
+                pendingResolve = resolve;
+            });
+        },
+        /**
+         * Return a Promise that will be resolved with a Uint8Array of at least the
+         * given length.  We buffer / accumulate received data until we have enough
+         * data.  Data is buffered even before you call this method, so be sure to
+         * explicitly wait for any and all data sent by the other side.
+         */
+        waitForDataWithAtLeastLength(length) {
+            if (pendingResolve) {
+                throw new Error("only one wait allowed at a time.");
+            }
+            if (receivedData && receivedData.length >= length) {
+                let promise = Promise.resolve(receivedData);
+                receivedData = null;
+                return promise;
+            }
+            console.log("(" + socketType + " waiting for " + length + " bytes)\n");
+            return new Promise(function(resolve, reject) {
+                pendingResolve = resolve;
+                wantDataLength = length;
+            });
+        },
+        waitForAnyDataAndClose() {
+            if (pendingResolve) {
+                throw new Error("only one wait allowed at a time.");
+            }
+
+            return new Promise(function(resolve, reject) {
+                pendingResolve = resolve;
+                // we may receive no data before getting close, in which case we want to
+                // return an empty array
+                receivedData = new Uint8Array();
+                wantDataAndClose = true;
+            });
+        },
+    };
+}
+
+var tcpsocket = class tcpsocket extends ExtensionAPI {
+    getAPI(context) {
+        return {
+            experiments: {
+                tcpsocket: {
+                    /** 
+                     * Send a DNS query stored in buf over a TCP socket to a 
+                     * nameserver addressed by addr
+                     */
+                    async sendDNSQuery(addr, buf) {
+                        let tcp_socket;
+                        let tcp_event_queue;
+                        let answer = new Uint8Array();
+                        try {
+                            tcp_socket = new TCPSocket(addr, 53, { binaryType: "arraybuffer" });
```

Comment:
For both the `tcpsocket` and `udpsocket` APIs, use `context.callOnClose` to register a cleanup handler to close sockets if the page unloads before the logic has completed. Examples: https://searchfox.org/mozilla-central/search?q=context.callOnClose&path=

Reply:
Roughly addressed by 7b2b7b97c6063f52b1cf68d925865e8a6205e44e. In `experiments.udpSockets.onDNSResponseReceived`, I think the `return () => ...` code that closes the sockets also runs when the extension closes, so I wonder I'm now duplicating code.